pipeline/train/train.py

```diff
@@ -0,0 +1,413 @@
+        logger.info(f" - artifacts: {self.artifacts}")
+        logger.info(f" - model_type: {self.model_type.value}")
+        logger.info(f" - teacher_mode: {self.teacher_mode.value}")
+        logger.info(f" - training_type: {self.training_type.value}")
+        logger.info(f" - best_model_metric: {self.best_model_metric}")
+        logger.info(f" - extra_marian_args: {self.extra_marian_args}")
+        logger.info(f" - marian_bin: {self.marian_bin}")
+        logger.info(f" - gpus: {self.gpus}")
+        logger.info(f" - workspace: {self.workspace}")
+        logger.info(f" - opustrainer_config: {self.opustrainer_config}")
+
+    def validate_args(self) -> None:
+        if self.extra_marian_args and self.extra_marian_args[0] != "--":
+            logger.error(" ".join(self.extra_marian_args))
+            raise Exception("Expected the extra marian args to be after a --")
+
+        # Validate input values.
+        if not self.vocab.exists():
+            raise Exception("Could not find the path to the vocab.")
+
+        if not self.marian_bin.exists():
+            raise Exception(f"Marian binary could not be found {self.marian_bin}")
+
+        for alignment_file in self.alignments_files:
+            if not alignment_file.exists():
+                raise Exception(f"Alignment file could not be found {alignment_file}")
+
+    def build_datasets(self) -> None:
+        # Start by building the training datasets, e.g.
+        #
+        #  corpus.enfr.tsv from:
+        #   - fetches/corpus.en.zst
+        #   - fetches/corpus.fr.zst
+        #   - fetches/corpus.aln.zst
+        #
+        #  mono.enfr.tsv from:
+        #   - fetches/mono.en.zst
+        #   - fetches/mono.fr.zst
+        #   - fetches/mono.aln.zst
+
+        for index, dataset_prefix in enumerate(self.train_set_prefixes):
+            alignments = None
+            if self.alignments_files:
+                alignments = self.alignments_files[index]
+                self.config_variables[f"dataset{index}"] = build_dataset_tsv(
+                    dataset_prefix, self.src, self.trg, alignments
+                )
+            else:
+                self.config_variables[f"dataset{index}"] = build_dataset_tsv(
+                    dataset_prefix, self.src, self.trg
+                )
+
+        # Then build out the validation set, for instance:
+        #
+        # devset.enfr.tsv from:
+        #  - fetches/devset.en.zst
+        #  - fetches/devset.fr.zst
+        self.validation_set = build_dataset_tsv(self.validation_set_prefix, self.src, self.trg)
+
+    def generate_opustrainer_config(
+        self,
+    ):
+        """
+        Generate an OpusTraininer config that points to the current datasets and language
+        options.
+        """
+        if self.model_type == ModelType.teacher:
+            teacher_mode = self.teacher_mode.value
+            if teacher_mode == TeacherMode.none.value:
+                logger.error("Teacher mode was not properly set")
```

Comment:
it should be `raise ValueError(...` if it's expected, otherwise there should be a default value in CLI args.